modules/moordyn/src/MoorDyn.f90

```diff
@@ -3885,27 +4068,39 @@ SUBROUTINE Init_Jacobian_x()
       !-----------------
       ! Free bodies
       DO l = 1,p%nFreeBodies                 ! Body m%BodyList(m%FreeBodyIs(l))
-         ! corresponds to state indices: (m%BodyStateIs1(l):m%BodyStateIs1(l)+5)
-         p%dx(idx+1:idx+3) = 0.1             ! body translational velocity [m/s]
-         p%dx(idx+4:idx+6) = 0.1             ! body rotational velocity [rad/s]
-         InitOut%LinNames_x(idx+1) = 'Body '//trim(num2lstr(m%FreeBodyIs(l)))//' Vx, m/s'
-         InitOut%LinNames_x(idx+2) = 'Body '//trim(num2lstr(m%FreeBodyIs(l)))//' Vy, m/s'
-         InitOut%LinNames_x(idx+3) = 'Body '//trim(num2lstr(m%FreeBodyIs(l)))//' Vz, m/s'
-         InitOut%LinNames_x(idx+4) = 'Body '//trim(num2lstr(m%FreeBodyIs(l)))//' omega_x, rad/s'
-         InitOut%LinNames_x(idx+5) = 'Body '//trim(num2lstr(m%FreeBodyIs(l)))//' omega_y, rad/s'
-         InitOut%LinNames_x(idx+6) = 'Body '//trim(num2lstr(m%FreeBodyIs(l)))//' omega_z, rad/s'
-         p%dxIdx_map2_xStateIdx(idx+1) = m%BodyStateIs1(l)+0         ! x%state index for Rx
-         p%dxIdx_map2_xStateIdx(idx+2) = m%BodyStateIs1(l)+1         ! x%state index for Ry
-         p%dxIdx_map2_xStateIdx(idx+3) = m%BodyStateIs1(l)+2         ! x%state index for Rz
-         p%dxIdx_map2_xStateIdx(idx+4) = m%BodyStateIs1(l)+3         ! x%state index for omega_x
-         p%dxIdx_map2_xStateIdx(idx+5) = m%BodyStateIs1(l)+4         ! x%state index for omega_y
-         p%dxIdx_map2_xStateIdx(idx+6) = m%BodyStateIs1(l)+5         ! x%state index for omega_z
-         idx = idx + 6
+         if (m%BodyList(m%FreeBodyIs(l))%typeNum == 2) then ! Coupled pinned body
+            ! corresponds to state indices: (m%BodyStateIs1(l):m%BodyStateIs1(l)+5)
```

Comment:
This may break linearization.  Did you resize the `p%dx` `InitOut%LinNames_x` accordingly?

We don't have a linearization case involving bodies though, so I don't have a good way to check.

Reply:
I believe so, those sizes are set by the size of the state vector which is determined when the input file is being read as the variable Nx: https://github.com/RyanDavies19/MoorDynF_ryan/blob/5ebcc5987650a7d6b1d7a53a8f942f32d1a41bd0/modules/moordyn/src/MoorDyn.f90#L881

For reading in the input file and allocating etc. I followed the exact same setup that was being used for the existing coupled pinned rods.

Reply:
Figured out my confusion.  In setting up the states there are 6 states for the pinned (3 DOF, and 3 derivatives of DOFs), and 12 states for the free bodies (6 DOF, and 6 derivatives of DOFs).  In the linearization section, we place motions first, so 3 states for pinned and 6 for free bodies. The derivatives are added to this matrix later giving the same total number of states, but ordered differently.